CONTRACTS: Avoid single conjunction in contracts instrumentation

feliperodri opened the pull request:
Currently, requires and ensures clauses are always combined into a single conjunction. The resulting clause is then used to create a single assertion or assumption, which makes really difficult to debug issues with large function contracts. This commit preserves the source location of the clauses by creating one assertion/assumption for each clause in the function contract (i.e., either requires or ensures).

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

feliperodri committed:
CONTRACTS: Avoid single conjunction in contracts instrumentation

Currently, requires and ensures clauses are always combined into
a single conjunction. The resulting clause is then used to create
a single assertion or assumption, which makes really difficult
to debug issues with large function contracts. This commit
preserves the source location of the clauses by creating one
assertion/assumption for each clause in the function contract
(i.e., either requires or ensures).

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

tautschnig commented on regression/contracts/assigns-replace-ignored-return-value/test.desc:
```diff
@@ -1,11 +1,13 @@
 ^VERIFICATION SUCCESSFUL$
 --
-^\[.*assigns.*\].*ignored_return_value.*FAILURE
+^\\[.*assigns.*\\].*ignored_return_value.*FAILURE
```
Why are you escaping the backslashes? I'm pretty sure this pattern will now not match any output, but will spuriously be doing this, and therefore make the test not meaningful anymore.

feliperodri replied:
This was a typo. Removed it.

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -772,35 +769,39 @@ void code_contractst::apply_function_contract(
-      ID_precondition);
-    is_fresh.update_requires(assertion);
-    new_program.destructive_append(assertion);
+    auto instantiated_clause =
+      to_lambda_expr(clause).application(instantiation_values);
+    if(!instantiated_clause.is_true())
+    {
+      if(
+        has_subexpr(instantiated_clause, ID_exists) ||
+        has_subexpr(instantiated_clause, ID_forall))
+        add_quantified_variable(instantiated_clause, mode);
+
+      goto_programt assertion;
+      converter.goto_convert(
+        code_assertt(instantiated_clause), assertion, mode);
+      assertion.instructions.back().source_location_nonconst() =
+        clause.source_location();
+      assertion.instructions.back().source_location_nonconst().set_line(
+        location.get_line());
+      assertion.instructions.back().source_location_nonconst().set_comment(
+        std::string("Check ")
+          .append(target_function.c_str())
+          .append("'s requires clause in ")
+          .append(function.c_str()));
+      assertion.instructions.back()
+        .source_location_nonconst()
+        .set_property_class(ID_precondition);
+      is_fresh.update_requires(assertion);
+      new_program.destructive_append(assertion);
+    }
```
1. Could all of this be moved into a function of its own? It's getting way too long.
2. I would advise against the "'s" and use "requires clause of function <name>" to avoid the oddity of using "'s" with a function the name of which ends in "s."

nwetzler replied:
I agree on both points.

feliperodri replied:
Implemented as suggested. I also refactored the initialization of history variables.

nwetzler replied:
Much cleaner now.

nwetzler commented on regression/contracts/assigns-replace-ignored-return-value/test.desc:
```diff
@@ -1,11 +1,13 @@
 ^VERIFICATION SUCCESSFUL$
 --
-^\[.*assigns.*\].*ignored_return_value.*FAILURE
+^\\[.*assigns.*\\].*ignored_return_value.*FAILURE
```
```suggestion
^\[.*assigns.*\].*ignored_return_value.*FAILURE
```

nwetzler commented on regression/contracts/variant_multidimensional_ackermann/test.desc:
```diff
@@ -1,8 +1,8 @@
-^\[ackermann.precondition\.\d+\] line 17 Check requires clause: SUCCESS$
-^\[ackermann.precondition\.\d+\] line 17 Check requires clause: SUCCESS$
+^\[ackermann.precondition\.\d+\] line 9 Check ackermann's requires clause in main: SUCCESS$
+^\[ackermann.precondition\.\d+\] line 34 Check ackermann's requires clause in ackermann: SUCCESS$
```
```suggestion
^\[ackermann.precondition\.\d+\] line \d+ Check ackermann's requires clause in ackermann: SUCCESS$
```
Perhaps abstract the line numbers above

nwetzler commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -772,35 +769,39 @@ void code_contractst::apply_function_contract(
+    if(!instantiated_clause.is_true())
+    {
+      if(
+        has_subexpr(instantiated_clause, ID_exists) ||
```
This formatting/indentation is weird.

feliperodri replied:
All the formatting/indentation is done by `clang-format`.

nwetzler commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -812,34 +813,31 @@ void code_contractst::apply_function_contract(
+    if(!instantiated_clause.is_false())
+    {
+      if(
+        has_subexpr(instantiated_clause, ID_exists) ||
```
Same formatting/indentation stuff here.

nwetzler commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -1445,67 +1440,67 @@ void code_contractst::add_contract_check(
+    if(!instantiated_clause.is_false())
+    {
+      if(
+        has_subexpr(instantiated_clause, ID_exists) ||
```
Formatting.

nwetzler reviewed (changes requested):
Looks pretty good.  Excited for this.

feliperodri committed:
CONTRACTS: Generator for contracts constraints

Creates common function to generate all contract constraint, i.e.,
assumptions and assertions based on requires and ensures clauses.

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

feliperodri committed:
CONTRACTS: Generator for history variables initialization

Creates common function to generate all instructions
to properly initialize history variables.

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

feliperodri committed:
CONTRACTS: Updates log message for requires clause

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -1445,67 +1451,43 @@ void code_contractst::add_contract_check(
-      code_assumet(requires), assumption, function_symbol.mode);
-    visitor.update_requires(assumption);
-    check.destructive_append(assumption);
+    if(!instantiated_clause.is_false())
+    {
```
If the clause happens to be false maybe we should throw a hard error

feliperodri replied:
Maybe yes. We should tell the user whether they are trying to assume a false assumption.

feliperodri replied:
At this stage, code like
```c
__CPROVER_requires(false)
__CPROVER_requires(0)
```
is always represented as a `notequal` exprt, i.e., `0 != 0`. So this if statement is never reached. Regardless, I added this check in cause we improve/change our front-end in the future.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -879,16 +877,27 @@ void code_contractst::apply_function_contract(
   {
-    is_fresh.update_ensures(ensures_pair.first);
-    new_program.destructive_append(ensures_pair.first);
+    if(!clause.is_false())
+    {
+      source_locationt _location = clause.source_location();
+      _location.set_comment("Assume ensures clause");
```
maybe throw a hard error if the clause is statically false ?

feliperodri replied:
See https://github.com/diffblue/cbmc/pull/7159#discussion_r984978712

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -771,36 +785,34 @@ void code_contractst::apply_function_contract(
+    auto instantiated_clause =
+      to_lambda_expr(clause).application(instantiation_values);
+
+    if(!instantiated_clause.is_true())
+    {
```
maybe we could add it regardless to keep the numbering of the clauses and the properties consistent ?

feliperodri replied:
Implemented as suggested.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -631,27 +631,44 @@ void code_contractst::replace_history_parameter(
-  // 2. instructions related to initializing the history variables
-  return std::make_pair(std::move(ensures_program), std::move(history));
+void generate_contract_constraints(
+  code_contractst &contract,
```
how about passing the converter directly instead of the whole `code_contractst` ?

feliperodri replied:
What about `add_quantified_variable`?

remi-delmas-3000 replied:
my bad I missed that in the first pass

remi-delmas-3000 reviewed (commented):
Similar to how `__CPROVER_assert(bool cond, char *msg)` accepts a string that shows up in the terminal output, 
do you think it would be possible to modify the front end to allow users to name the clauses with an optional string ?

i.e:

```c
int foo(int *arr, size_t size)
__CPROVER_requires(size > 0 , "some text")
__CPROVER_ensures(__CPROVER_return_value == 1 , "some text")
{
}
```

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -771,36 +785,34 @@ void code_contractst::apply_function_contract(
+    {
+      source_locationt _location = clause.source_location();
+      _location.set_line(location.get_line());
+      _location.set_comment(std::string("Check requires clause of ")
```
could we also embed the string representation of the formula being asserted in the source location ?

feliperodri replied:
See https://github.com/diffblue/cbmc/pull/7159#discussion_r984978914

nwetzler commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -631,27 +631,44 @@ void code_contractst::replace_history_parameter(
-  // 1. instructions corresponding to the ensures clause
-  // 2. instructions related to initializing the history variables
-  return std::make_pair(std::move(ensures_program), std::move(history));
+void generate_contract_constraints(
```
Does this function need a declaration in the header file?  Also, can you provide a description in comments?

feliperodri replied:
This should be a static function (good catch), thus we don't need it in the header file. I also added a small description for it.

feliperodri commented:
> Similar to how `__CPROVER_assert(bool cond, char *msg)` accepts a string that shows up in the terminal output, do you think it would be possible to modify the front end to allow users to name the clauses with an optional string ?
> 
> i.e:
> 
> ```c
> int foo(int *arr, size_t size)
> __CPROVER_requires(size > 0 , "some text")
> __CPROVER_ensures(__CPROVER_return_value == 1 , "some text")
> {
> }
> ```

@remi-delmas-3000 certainly, but it should go on a separate PR. This change is orthogonal to this change. I'll create a GitHub issue to track this request.

feliperodri commented:
@remi-delmas-3000 we can track the request for requires and ensures clauses with descriptions at https://github.com/diffblue/cbmc/issues/7181.

jimgrundy commented:
Thanks for addressing this issue @feliperodri. It's long been on our wish list.

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -631,27 +631,46 @@ void code_contractst::replace_history_parameter(
+  if(
+    has_subexpr(instantiated_clause, ID_exists) ||
+    has_subexpr(instantiated_clause, ID_forall))
+    contract.add_quantified_variable(instantiated_clause, mode);
```
Nit pick: please use braces around the body when the condition or the body span multiple lines.

feliperodri replied:
Done.

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -631,27 +631,46 @@ void code_contractst::replace_history_parameter(
+      code_assumet(instantiated_clause), constraint, mode);
+  else
+    contract.get_converter().goto_convert(
+      code_assertt(instantiated_clause), constraint, mode);
```
As above: braces, please.

feliperodri replied:
Done!

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -631,27 +631,46 @@ void code_contractst::replace_history_parameter(
+    contract.add_quantified_variable(instantiated_clause, mode);
+
+  goto_programt constraint;
+  if((location.get_property_class() == ID_assume))
```
Nit pick: nested parentheses?

feliperodri replied:
fixed.

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -631,27 +631,46 @@ void code_contractst::replace_history_parameter(
+  exprt &instantiated_clause,
+  const irep_idt &mode,
+  is_fresh_baset &is_fresh,
+  void (is_fresh_baset::*is_fresh_update)(goto_programt &),
```
I'd prefer to see use of `std::function` instead of having to pass two arguments, one of which is a pointer, just to access a member function.

feliperodri replied:
Good catch! Implemented as suggested.

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -810,36 +820,22 @@ void code_contractst::apply_function_contract(
-
-    auto assumption = code_assumet(ensures);
-    ensures_pair =
-      create_ensures_instruction(assumption, ensures.source_location(), mode);
-
-    // add all the history variable initialization instructions
-    new_program.destructive_append(ensures_pair.second);
+    auto instantiated_clause =
+      to_lambda_expr(clause).application(instantiation_values);
+    instantiated_clause.add_source_location() = clause.source_location();
+    generate_history_variables_initialization(
+      instantiated_clause, mode, new_program);
+    instantiated_ensures_clauses.push_back(instantiated_clause);
```
Would it be possible to delay this piece of code until we actually use `instantiated_ensures_clauses`?

feliperodri replied:
@tautschnig I'm storing all instantiated clauses in `instantiated_ensures_clauses`, so I don't have to call `to_lambda_expr(clause).application(instantiation_values)` twice. I need to instantiate them before adding instructions for history variables, which must be done here (in that order). I'll use them later when I add the instructions for ensures clauses. So, I either keep the code that way or I remove `instantiated_ensures_clauses` and call the `application` method twice.
